modules/images/image-loading-optimization/storage/post-type.php

```diff
@@ -0,0 +1,181 @@
+}
+
+const ILO_PAGE_METRICS_POST_TYPE = 'ilo_page_metrics';
+
+/**
+ * Register post type for page metrics storage.
+ *
+ * This the configuration for this post type is similar to the oembed_cache in core.
+ */
+function ilo_register_page_metrics_post_type() /*: void (in PHP 7.1) */ {
+	register_post_type(
+		ILO_PAGE_METRICS_POST_TYPE,
+		array(
+			'labels'           => array(
+				'name'          => __( 'Page Metrics', 'performance-lab' ),
+				'singular_name' => __( 'Page Metrics', 'performance-lab' ),
+			),
+			'public'           => false,
+			'hierarchical'     => false,
+			'rewrite'          => false,
+			'query_var'        => false,
+			'delete_with_user' => false,
+			'can_export'       => false,
+			'supports'         => array( 'title' ), // The original URL is stored in the post_title, and the post_name is a hash of the query vars.
+		)
+	);
+}
+add_action( 'init', 'ilo_register_page_metrics_post_type' );
+
+/**
+ * Get page metrics post.
+ *
+ * @param string $slug Page metrics slug.
+ * @return WP_Post|null Post object if exists.
+ */
+function ilo_get_page_metrics_post( string $slug ) /*: ?WP_Post (in PHP 7.1) */ {
+	$post_query = new WP_Query(
+		array(
+			'post_type'              => ILO_PAGE_METRICS_POST_TYPE,
+			'post_status'            => 'publish',
+			'name'                   => $slug,
+			'posts_per_page'         => 1,
+			'no_found_rows'          => true,
+			'cache_results'          => true,
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'lazy_load_term_meta'    => false,
+		)
+	);
+
+	$post = current( $post_query->posts );
+	if ( $post instanceof WP_Post ) {
+		return $post;
+	} else {
+		return null;
+	}
+}
+
+/**
+ * Parses post content in page metrics post.
+ *
+ * @param WP_Post $post Page metrics post.
+ * @return array|WP_Error Page metrics when valid, or WP_Error otherwise.
+ */
+function ilo_parse_stored_page_metrics( WP_Post $post ) /*: array|WP_Error (in PHP 8) */ {
+	$page_metrics = json_decode( $post->post_content, true );
+	if ( json_last_error() ) {
+		return new WP_Error(
+			'page_metrics_json_parse_error',
+			sprintf(
+				/* translators: 1: Post type slug, 2: JSON error message */
+				__( 'Contents of %1$s post type not valid JSON: %2$s', 'performance-lab' ),
+				ILO_PAGE_METRICS_POST_TYPE,
+				json_last_error_msg()
+			)
+		);
+	}
+	if ( ! is_array( $page_metrics ) ) {
+		return new WP_Error(
+			'page_metrics_invalid_data_format',
+			sprintf(
+				/* translators: %s is post type slug */
+				__( 'Contents of %s post type was not a JSON array.', 'performance-lab' ),
+				ILO_PAGE_METRICS_POST_TYPE
+			)
+		);
+	}
+	return $page_metrics;
+}
+
+/**
+ * Gets page metrics for a slug.
+ *
+ * This is a convenience abstractions for lower-level functions.
+ *
+ * @see ilo_get_page_metrics_post()
+ * @see ilo_parse_stored_page_metrics()
+ *
+ * @param string $slug Page metrics slug.
+ * @return array Page metrics data, or empty array if invalid.
```

Comment:
curious why you decided to return an empty array here, I would expect an error

Reply:
I've reworked this in f566663. Now any errors are emitted from `ilo_parse_stored_page_metrics()` via `wp_trigger_error()` rather than being returned in a `WP_Error`. Since the parse error case should be truly exceptional, and the error recovery case is to just blow the data away and start again with an empty array, I think this makes more sense.